src/com/sun/jna/Function.java

```diff
@@ -620,6 +620,8 @@ private Object convertArgument(Object[] args, int index,
         } else if (argClass.isArray()){
             throw new IllegalArgumentException("Unsupported array argument type: "
                                                + argClass.getComponentType());
+        } else if (JString.class == argClass) {
+            return arg.toString();
```

Comment:
An alternative here would be to avoid `String` or `WString` conversion when allow_objects is set, and require and explicit `NativeString` if that's really what you want.  That avoids the addition of a new possibly confusing class which is used in a minority of cases.